File: leetcode/math/2524_Maximum_Frequency_Score_of_a_Subarray/codes/math.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MOD 1000000007
#define MAX_VAL 1000000

typedef long long i64;

int modPlus(int a, int b) {
    return ((i64)a + b) % MOD;
}

int modMinus(int a, int b) {
    return ((i64)a - b + MOD) % MOD;
}

int modMult(int a, int b) {
    return ((i64)a * b) % MOD;
}

int modPow(int base, int exp) {
    i64 result = 1;
    while (exp > 0) {
        if (exp & 1) result = modMult(result, base);
        exp >>= 1;
        base = modMult(base, base);
    }
    return result;
}

int maxFrequencyScore(int* nums, int n, int k) {
    int* freq = (int*)calloc(MAX_VAL + 1, sizeof(int));
    int result = 0, sum = 0;

    for (int r = 0; r < n; ++r) {
        int val = nums[r];
        int prevExp = freq[val]++;
        if (prevExp > 0) {
            int contribution = modPow(val, prevExp);
            sum = modMinus(sum, contribution);
            contribution = modMult(contribution, val);
            sum = modPlus(sum, contribution);
        } else {
            sum = modPlus(sum, val);
        }

        if (r >= k) {
            int leftVal = nums[r - k];
            int nextExp = --freq[leftVal];
            if (nextExp > 0) {
                int contribution = modPow(leftVal, nextExp + 1);
                sum = modMinus(sum, contribution);
                contribution = modPow(leftVal, nextExp);
                sum = modPlus(sum, contribution);
            } else {
                sum = modMinus(sum, leftVal);
            }
        }

        if (r >= k - 1) {
            if (sum > result) result = sum;
        }
    }

    free(freq);
    return result;
}

File: leetcode/math/2524_Maximum_Frequency_Score_of_a_Subarray/codes/math.py
MOD = int(1e9 + 7)
MAX_VAL = int(1e6)

def modPlus(a, b, mod=MOD):
    return (a + b) % mod

def modMinus(a, b, mod=MOD):
    return (a - b + mod) % mod

def modMult(a, b, mod=MOD):
    return (a * b) % mod

def modPow(base, exp, mod=MOD):
    result = 1
    while exp:
        if exp & 1:
            result = modMult(result, base, mod)
        base = modMult(base, base, mod)
        exp >>= 1
    return result

class Solution:
    def maxFrequencyScore(self, nums: List[int], k: int) -> int:
        from collections import defaultdict

        freq = defaultdict(int)
        n = len(nums)
        result = 0
        total = 0

        for r in range(n):
            prevExp = freq[nums[r]]
            freq[nums[r]] += 1

            if prevExp > 0:
                contribution = modPow(nums[r], prevExp)
                total = modMinus(total, contribution)
                contribution = modMult(contribution, nums[r])
                total = modPlus(total, contribution)
            else:
                total = modPlus(total, nums[r])

            if r >= k:
                left = nums[r - k]
                nextExp = freq[left] - 1
                freq[left] -= 1

                if nextExp > 0:
                    contribution = modPow(left, nextExp + 1)
                    total = modMinus(total, contribution)
                    contribution = modPow(left, nextExp)
                    total = modPlus(total, contribution)
                else:
                    total = modMinus(total, left)

            if r >= k - 1:
                result = max(result, total)

        return result
